### PR TITLE
Port paper ulyanov_et_al_2016

### PR DESCRIPTION
**Pull request by jbueltemeier**

### CODE REVIEW

File: pystiche_papers/ulyanov_et_al_2016/core.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    num_epochs: Optional[int] = None,
+    get_optimizer: Optional[Callable[[UlyanovEtAl2016Transformer], Optimizer]] = None,
+    quiet: bool = False,
+    logger: Optional[OptimLogger] = None,
+    log_fn: Optional[
+        Callable[[int, Union[torch.Tensor, pystiche.LossDict], float, float], None]
+    ] = None,
+) -> nn.Module:
+    if isinstance(style, str):
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (pmeier) [inline pystiche_papers/ulyanov_et_al_2016/core.py:60]: We can use `get_device` from `pystiche.misc` here 
```suggestion
        device = get_device()
```

File: pystiche_papers/ulyanov_et_al_2016/loss.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        input_repr: torch.Tensor,
+        target_repr: torch.Tensor,
+        ctx: Optional[torch.Tensor],
+    ) -> torch.Tensor:
+        score = super().calculate_score(input_repr, target_repr, ctx)
+        if not self.double_batch_size_mean:
+            return score
+        else:
+            batch_size = extract_batch_size(input_repr)
+            return score / batch_size
```
Comment (pmeier) [inline pystiche_papers/ulyanov_et_al_2016/loss.py:43]:
```suggestion

        batch_size = extract_batch_size(input_repr)
        return score / batch_size
```

File: pystiche_papers/ulyanov_et_al_2016/loss.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        input_repr: torch.Tensor,
+        target_repr: torch.Tensor,
+        ctx: Optional[torch.Tensor],
+    ) -> torch.Tensor:
+        score = super().calculate_score(input_repr, target_repr, ctx)
+        if not self.double_batch_size_mean:
+            return score
+        else:
+            batch_size = input_repr.size()[0]
+            return score / batch_size
```
Comment (pmeier) [inline pystiche_papers/ulyanov_et_al_2016/loss.py:95]:
```suggestion

        batch_size = input_repr.size()[0]
        return score / batch_size
```

File: pystiche_papers/ulyanov_et_al_2016/utils.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+from torch import nn, optim
+from torch.optim.lr_scheduler import ExponentialLR
+from torch.optim.optimizer import Optimizer
+
+from pystiche.enc import MultiLayerEncoder, vgg19_multi_layer_encoder
+from pystiche.image.transforms.processing import CaffePostprocessing, CaffePreprocessing
+
+
+def ulyanov_et_al_2016_multi_layer_encoder() -> MultiLayerEncoder:
```
Comment (pmeier) [inline pystiche_papers/ulyanov_et_al_2016/utils.py:11]: Make this more specific

```suggestion
def ulyanov_et_al_2016_multi_layer_encoder() -> VGGMultiLayerEncoder:
```

File: pystiche_papers/ulyanov_et_al_2016/utils.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    return CaffePreprocessing()
+
+
+def ulyanov_et_al_2016_postprocessor() -> CaffePostprocessing:
+    return CaffePostprocessing()
+
+
+def ulyanov_et_al_2016_optimizer(
+    transformer: nn.Module, impl_params: bool = True, instance_norm: bool = True
+) -> Optimizer:
```
Comment (pmeier) [inline pystiche_papers/ulyanov_et_al_2016/utils.py:27]: Make this more specific
```suggestion
) -> optim.Adam:
```

File: pystiche_papers/ulyanov_et_al_2016/utils.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        exp = self.last_epoch - self.delay + 1
+        if exp > 0:
+            return [base_lr * self.gamma ** exp for base_lr in self.base_lrs]
+        else:
+            return self.base_lrs
+
+
+def ulyanov_et_al_2016_lr_scheduler(
+    optimizer: Optimizer, impl_params: bool = True,
+) -> Optional[ExponentialLR]:
```
Comment (pmeier) [inline pystiche_papers/ulyanov_et_al_2016/utils.py:56]: Why is this `Optional`? We won't return `None`.

File: replication/ulyanov_et_al_2016/main.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+from pystiche_papers.utils import save_state_dict
+
+
+def training_texture(args):
+    textures = (
+        "cezanne",
+        "bricks",
+        "pebble",
+        "pixels",
+        # "peppers",
```
Comment (pmeier) [inline replication/ulyanov_et_al_2016/main.py:26]: What is this about?

File: replication/ulyanov_et_al_2016/main.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    )
+
+    dataset = ulyanov_et_al_2016_dataset(
+        path.join(args.dataset_dir, "texture"),
+        impl_params=args.impl_params,
+        instance_norm=args.instance_norm,
+    )  # FIXME: right dataset -> random from Imagenet (current examples from caltech256)
```
Comment (pmeier) [inline replication/ulyanov_et_al_2016/main.py:33]: Isn't that irrelevant from the implementation standpoint if we are using an `ImageFolderDataset`? Nevertheless, we should document this later on.

File: replication/ulyanov_et_al_2016/main.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+    image_loader = ulyanov_et_al_2016_image_loader(
+        dataset,
+        impl_params=args.impl_params,
+        instance_norm=args.instance_norm,
+        stylization=args.stylization,
+        pin_memory=str(args.device).startswith("cuda"),
+    )
+
+    images = ulyanov_et_al_2016_images()
+    images.download(root=args.image_source_dir)
```
Comment (pmeier) [inline replication/ulyanov_et_al_2016/main.py:43]:
```suggestion
    images.download(args.image_source_dir)
```

File: replication/ulyanov_et_al_2016/main.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+        write_image(output_image, output_file)
+
+
+def training_style(args):
+    contents = (
+        "karya",
+        "tiger",
+        "neckarfront",
+        # "che_high",
+        # "the_tower_of_babel"
```
Comment (pmeier) [inline replication/ulyanov_et_al_2016/main.py:104]: What about them?

File: replication/ulyanov_et_al_2016/main.py [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+        "pleades",
+        "starry",
+        "turner",
+    )
+
+    dataset = ulyanov_et_al_2016_dataset(
+        path.join(args.dataset_dir, "style"),
+        impl_params=args.impl_params,
+        instance_norm=args.instance_norm,
+    )  # FIXME: right dataset -> random from Imagenet (current examples from caltech256)
```
Comment (pmeier) [inline replication/ulyanov_et_al_2016/main.py:122]: See above.

File: replication/ulyanov_et_al_2016/main.py [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+    image_loader = ulyanov_et_al_2016_image_loader(
+        dataset,
+        impl_params=args.impl_params,
+        instance_norm=args.instance_norm,
+        stylization=args.stylization,
+        pin_memory=str(args.device).startswith("cuda"),
+    )
+
+    images = ulyanov_et_al_2016_images()
+    images.download(root=args.image_source_dir)
```
Comment (pmeier) [inline replication/ulyanov_et_al_2016/main.py:132]:
```suggestion
    images.download(args.image_source_dir)
```

File: replication/ulyanov_et_al_2016/main.py [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+
+    if model_dir is None:
+        model_dir = path.join(here, "data", "models")
+    model_dir = process_dir(model_dir)
+
+    if device is None:
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+    if isinstance(device, str):
+        device = torch.device(device)
+
```
Comment (pmeier) [inline replication/ulyanov_et_al_2016/main.py:213]: See above.

```suggestion
    device = get_device()
```

File: replication/ulyanov_et_al_2016/main.py [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+        logger=logger,
+        quiet=quiet,
+    )
+
+
+if __name__ == "__main__":
+    args = parse_input()
+    # args.stylization = False
+    # training_texture(args)
+    args.stylization = True
```
Comment (pmeier) [inline replication/ulyanov_et_al_2016/main.py:234]:
```suggestion
```

**Comment by pmeier** [comment]
@jbueltemeier the `error: unused 'type: ignore comment` in check `lint / static_typing` are raised by `mypy`, since the underlying issue was fixed in pmeier/pystiche/pull/317. Could you send a PR fixing them in the existing code base and afterwards rebase this PR on that? If you have trouble to reproduce these errors locally, re-install `pystiche` to get the latest version.
